File: it-agents/skills/code-architect/references/code-examples/csharp/Command.cs
// Command — encapsulate a request as an object; supports undo.
using System;
using System.Collections.Generic;

namespace Patterns.Command
{
    // Receiver holds the state the commands act upon.
    sealed class Receiver
    {
        public int Value { get; private set; }
        public void Apply(int delta) => Value += delta;
    }

    interface ICommand
    {
        void Execute();
        void Undo();
    }

    sealed class AddCommand : ICommand
    {
        private readonly Receiver _receiver;
        private readonly int _amount;
        public AddCommand(Receiver receiver, int amount)
        {
            _receiver = receiver;
            _amount = amount;
        }
        public void Execute() => _receiver.Apply(_amount);
        public void Undo() => _receiver.Apply(-_amount);
    }

    // Invoker keeps a history to support undo.
    sealed class Invoker
    {
        private readonly Stack<ICommand> _history = new();
        public void Run(ICommand command)
        {
            command.Execute();
            _history.Push(command);
        }
        public void Undo()
        {
            if (_history.Count > 0) _history.Pop().Undo();
        }
    }

    static class Demo
    {
        public static void Run()
        {
            var receiver = new Receiver();
            var invoker = new Invoker();
            invoker.Run(new AddCommand(receiver, 5));
            invoker.Run(new AddCommand(receiver, 3));
            Console.WriteLine($"After commands: {receiver.Value}");
            invoker.Undo();
            Console.WriteLine($"After undo: {receiver.Value}");
        }

        public static void Main() => Run();
    }
}

File: it-agents/skills/code-architect/references/code-examples/go/command.go
// Command: encapsulate a request as an object, enabling queuing, logging
// and undo of operations.
package main

import "fmt"

// receiver holds the state that commands act upon.
type receiver struct {
	value int
}

// Command is an action plus its inverse.
type Command interface {
	Execute()
	Undo()
}

type addCommand struct {
	target *receiver
	amount int
}

func (c addCommand) Execute() { c.target.value += c.amount }
func (c addCommand) Undo()    { c.target.value -= c.amount }

// invoker runs commands and keeps history for undo.
type invoker struct {
	history []Command
}

func (i *invoker) Run(c Command) {
	c.Execute()
	i.history = append(i.history, c)
}

func (i *invoker) UndoLast() {
	if n := len(i.history); n > 0 {
		i.history[n-1].Undo()
		i.history = i.history[:n-1]
	}
}

func main() {
	r := &receiver{}
	inv := &invoker{}

	inv.Run(addCommand{target: r, amount: 5})
	inv.Run(addCommand{target: r, amount: 3})
	fmt.Println("after run:", r.value)

	inv.UndoLast()
	fmt.Println("after undo:", r.value)
}
